SwiftCompilerSources/Sources/Optimizer/FunctionPasses/StackPromotion.swift

```diff
@@ -200,59 +195,83 @@ func tryPromoteAlloc(_ allocRef: AllocRefInstBase,
 //                              utility functions
 //===----------------------------------------------------------------------===//
 
-private extension EscapeInfo {
-  mutating func getDominatingBlockOfAllUsePoints(_ value: SingleValueInstruction,
-                                                 domTree: DominatorTree) -> BasicBlock {
-    // The starting point
-    var dominatingBlock = value.block
-
-    _ = isEscaping(object: value, visitUse: { op, _, _ in
-        let defBlock = op.value.definingBlock
-        if defBlock.dominates(dominatingBlock, domTree) {
-          dominatingBlock = defBlock
-        }
-        return .continueWalking
-      })
-    return dominatingBlock
+private func getDominatingBlockOfAllUsePoints(context: PassContext,
+                                              _ value: SingleValueInstruction,
+                                              domTree: DominatorTree) -> BasicBlock {
+  struct Visitor : EscapeInfoWalkerVisitor {
+    var dominatingBlock: BasicBlock
+    let domTree: DominatorTree
+    mutating func visitUse(operand: Operand, path: Path, state: State) -> UseVisitResult {
+      let defBlock = operand.value.definingBlock
+      if defBlock.dominates(dominatingBlock, domTree) {
+        dominatingBlock = defBlock
+      }
+      return .continueWalk
+    }
   }
+  
+  var walker = EscapeInfoWalker(calleeAnalysis: context.calleeAnalysis, visitor: Visitor(dominatingBlock: value.block, domTree: domTree))
+  _ = walker.isEscaping(object: value)
+  return walker.visitor.dominatingBlock
+}
 
+struct InnerAndOuterLiferanges {
```

Comment:
There is no need to make this a struct. You can just put the Visitor into the `of` function. And make it a top-level function (instead of static) and rename it to `computeInnerAndOuterLiferanges` (as it was before).

Also please add a comment with a warning that the returned liferanges have to be `deinitialize`d by the caller - until they are true move-only types.